File: num_0_100/100_SameTree.py
# -*- coding:utf-8 -*-
# Created data: 2020906

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSameTree(self, p: TreeNode, q: TreeNode) -> bool:
        if not p and not q:
            return True
        elif not p or not q:
            return False
        elif p.val != q.val:
            return False
        else:
            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

# 广度优先遍历的核心是用队列,node先先进先出
class buildTree(object):
    def __init__(self):
        self.root = None
    # 这是我写过最繁琐的二叉树的构建
    # 时间复杂度不好,以后不打算这么写了
    def add_node(self, ele):
        queue = [self.root]
        while queue:
            node = queue.pop(0)
            if node.left is None:
                node.left = TreeNode(ele)
                return
            elif node.right is None:
                node.right = TreeNode(ele)
                return
            queue.append(node.left)
            queue.append(node.right)

    def establish(self, ls):
        for ele in ls:
            if self.root is None:
                self.root = TreeNode(ele)
            self.add_node(ele)
        return self.root

    def establish_abd(self, ls: list):
        if not ls: return None
        queue = []
        for ele in ls:
            if not self.root:
                self.root = TreeNode(ele)
                queue.append(self.root)
                continue
            while queue:
                node = queue.pop(0)
                if node.left is None:
                    node.left = TreeNode(ele)
                    queue.insert(0, node)
                    break
                elif node.right is None:
                    node.right = TreeNode(ele)
                    queue.append(node.left)
                    queue.append(node.right)
                    break
                else:
                    pass
        return self.root


def breath_travel(root):
    if root is None:
        print(None); return
    queue = [root]
    while queue:
        node = queue.pop(0)
        print(node.val)
        if node.left:
            queue.append(node.left)
        if node.right:
            queue.append(node.right)

if __name__ == "__main__":
    ls1 = [i for i in range(1, 16)]
    ls2 = [i for i in range(1, 15)]
    tree_root1 = buildTree().establish(ls1)
    tree_root2 = buildTree().establish(ls2)
    # breath_travel(tree_root1)
    # breath_travel(tree_root2)
    solu = Solution().isSameTree(tree_root1, tree_root2)
    print(solu)

File: num_0_100/100_SameTree.cpp
#include<iostream>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

namespace sametreespace {

    class Solution {
    public:
        Solution(){
        srand(time(NULL));
        }
        bool isSameTree(TreeNode *p, TreeNode *q) {
    //        if (random()%2){
        if (1){
            return isSameTree1(p, q);
        }
    //        return isSameTree2(p, q);
        }

        bool isSameTree1(TreeNode *p, TreeNode *q) {
        if(!p && !q) return true;
        if(!p || !q) return false;
        if(p->val != q->val) return false;
        return (isSameTree(p->left, q->left) &&
            isSameTree(p->right, q->right));
        }
    //    bool isSameTree2(TreeNode *p, TreeNode *q) {
    //        queue<TreeNode*> q1, q2;
    //        q1.push(p);
    //        q2.push(q);
    //        while (q1.size()>0 && q2.size()>0 ){
    //            TreeNode* p1 = q1.front();
    //            q1.pop();
    //            TreeNode* p2 = q2.front();
    //            q2.pop();
    //            if (!p1 && !p2) continue;
    //            if (!p1 || !p2)  return false;
    //            if ( p1->val != p2->val) {
    //                return false;
    //            }
    //            q1.push(p1->left);
    //            q2.push(p2->left);
    //            q1.push(p1->right);
    //            q2.push(p2->right);
    //        }
    //        return true;
    //    }
    };
};


void sameTreeMain(void){
    TreeNode root1(1);
    TreeNode root2(1);
    TreeNode t2(2);
    TreeNode t3(3);
    TreeNode t4(4);
    TreeNode t5(5);

    root1.left = &t2;
    root1.right = &t3;
    root2.left = &t2;
    root2.right = &t3;
    t2.left = &t4;
    t2.right = &t5;
    sametreespace::Solution solu;
    std::cout << solu.isSameTree(&root1, &root2) << std::endl;
}
